expcore/auto_play.lua

```diff
@@ -0,0 +1,519 @@
+        if selector == element then
+            category = selector_category
+        end
+    end
+    if not category then return end
+
+    -- Update vote counts displayed to everyone
+    local list_items = vote_selector_list(category)
+    for _, connected_player in pairs(game.connected_players) do
+        local selector = player_data[connected_player.index].vote_selectors[category]
+        if selector and selector.valid then
+            selector.items = list_items
+        end
+    end
+end)
+
+
+--- Welcome screen presenting users with a choice of joining the game or
+-- spectating in the game.
+local RoleSelectorWindow =
+Gui.element(function(_, parent)
+    local outer_frame =
+    parent.add{
+        name = "role_selector_window",
+        type = 'frame',
+        direction = 'vertical',
+        caption = "Factorio Car Race",
+    }
+    outer_frame.style.use_header_filler = false
+
+    local inner_frame =
+    outer_frame.add{
+        name = "role_selector_inner_frame",
+        type = 'frame',
+        direction = 'vertical',
+        style = 'inside_shallow_frame_with_padding',
+    }
+    inner_frame.style.width = 250 + 24
+
+    local sprite =
+    inner_frame.add{
+        type = 'sprite',
+        sprite = 'file/modules/gui/race.png',
+    }
+    sprite.style.stretch_image_to_widget_size = true
+    sprite.style.width = 250
+    sprite.style.height = 167
+    sprite.style.margin = {0, 0, 4, 0}
+
+    local label =
+    inner_frame.add{
+        type = 'label',
+        caption = "Be the first to the finish line in this multiplayer car race mini-game.",
+    }
+    label.style.single_line = false
+
+
+    local dialog_row =
+    outer_frame.add{
+        type = 'flow',
+        direction = 'horizontal',
+        style = 'dialog_buttons_horizontal_flow'
+    }
+
+    SpectateButton(dialog_row)
+
+    local pusher =
+    dialog_row.add{
+        type = 'empty-widget',
+    }
+    pusher.style.horizontally_stretchable = true
+    pusher.style.height = 32
+
+    JoinButton(dialog_row)
+
+    return outer_frame
+end)
+
+
+--- Race voting window.  Presents vote selectors for race map, fuel and lap count, and
+-- a button to toggle participation in the game.
+local RaceVoteWindow =
+Gui.element(function(_, parent, data)
+    local outer_frame =
+    parent.add{
+        name = "race_vote_window",
+        type = 'frame',
+        direction = 'vertical',
+        caption = "...",
+    }
+    outer_frame.style.use_header_filler = true
+    outer_frame.location = {400, 100}
+
+    local inner_frame =
+    outer_frame.add{
+        name = "race_vote_inner_frame",
+        type = 'frame',
+        direction = 'vertical',
+        style = 'inside_shallow_frame',
+    }
+
+    local header =
+    inner_frame.add{
+        type = 'frame',
+        direction = 'horizontal',
+        caption = "Vote for map",
+        style = 'subheader_frame',
+    }
+    header.style.padding = {4, 8}
+    header.style.use_header_filler = false
```

Comment:
This can be reduced to:
```lua
local header = Gui.header(inner_frame, "Vote for map")
header.style.padding = { 4, 8 }
```
(only difference is the caption is added as a label element)

However, I think it would be better to have the title of the frame be "Vote for map" rather than using a header.
The "Spectating" vs "Joining" can be done with only a label as you have done for the start label.

Reply:
I think that can be changed on in a later PR.

Reply:
The first half of the suggestion has been addressed so this is now resolved.
I am happy for the other half to be included in a later PR.